Car_owners/tests/test_views.py

```diff
@@ -0,0 +1,269 @@
+
+    # try to get that car
+    response_get_car = api_client.get(f"/app/cars/{car_id}/", format="json")
+    assert response_get_car.status_code == status.HTTP_404_NOT_FOUND
+
+    # Car does not exist
+    response_delete_car = api_client.delete(f"/app/cars/{car_id + 1}/", format="json")
+    assert response_delete_car.status_code == status.HTTP_404_NOT_FOUND
+
+
+@pytest.mark.parametrize(
+    ("data", "expected_message"),
+    [
+        ({"phone": "asd1234cvb"}, "Phone number can contain only digits"),
+        ({"phone": "12345678911111"}, "Phone number is too long"),
+        ({"phone": "123"}, "Phone number is too short"),
+        ({"name": "@#$"}, "name can contain only letters and '-' without whitespaces"),
+        (
+            {"surname": "@#$"},
+            "surname can contain only letters and '-' without whitespaces",
+        ),
+        (
+            {"ordering": "invalid ordering"},
+            f"Ordering should be one of the following: ",
+        ),
+    ],
+)
+@pytest.mark.django_db
+def test_owner_request_validation(
+    api_client: APIClient, data: dict[str, str], expected_message: str
+) -> None:
+    ((key, value),) = data.items()
+    if key == "ordering":
+        owner_view = OwnerViewSet()
+        ord_fields_string = ", ".join(owner_view.ordering_fields)
+        expected_message += ord_fields_string
+
+    response_get_owner_invalid_data = api_client.get(
+        "/app/owners/", data=data, format="json"
+    )
+    assert response_get_owner_invalid_data.status_code == status.HTTP_400_BAD_REQUEST
+    assert response_get_owner_invalid_data.data[key] == expected_message
+
+
+@pytest.mark.parametrize(
+    ("model_data", "model_str"),
+    [
+        ({"name": "Andrzej", "surname": "Starczyk", "phone": "575849675"}, "owner"),
+        (
+            {
+                "brand": "Ford",
+                "model": "Mondeo",
+                "production_date": "2023-01-01",
+            },
+            "car",
+        ),
+    ],
+)
+@pytest.mark.django_db
+def test_response_no_object_found(
+    api_client: APIClient, model_data: dict[str, str], model_str: str
+) -> None:
+    response_get_model = api_client.get(
+        f"/app/{model_str}s/", data=model_data, format="json"
+    )
+    assert response_get_model.status_code == status.HTTP_200_OK
+    assert response_get_model.data == f"There is no {model_str.title()} with given data"
+
+
+@pytest.mark.parametrize(
+    ("data", "expected_message"),
+    [
+        (
+            {"production_date": datetime.date.today() + datetime.timedelta(days=1)},
+            "Production date cannot be from the future.",
+        ),
+        ({"ordering": "invalid ordering"}, "Ordering should be one of the following: "),
+    ],
+)
+@pytest.mark.django_db
+def test_car_request_validation(
+    api_client: APIClient, data: dict[str, datetime.date | str], expected_message: str
+) -> None:
+    ((key, value),) = data.items()
+    if key == "ordering":
```

Comment:
looks like a separate test

Reply:
It's not separate test - the logic for "ordering" is expanding 'expected_message', because it depends on ordering fields from ViewSet. I didn't want to hardcode it, because fields, which allow ordering may change in the future. With that logic, test will be still valid, no matter what fields will allow ordering.

Reply:
I think testing ordering is another test case, so here you can just choose one option and proceed with stuff you actually want to test within this test

Reply:
Current version is fine :+1: